src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,79 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, vect, stride, offset)
+    for i in offset:stride:length(vect)
+        push!(col, vect[i])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    if !(indexcols isa Symbol || allunique(indexcols))
+        throw(ArgumentError("Elements of $indexcols must be unique"))
+    end
+    sum(nonunique(df, indexcols))>0 && @warn "duplicate rows in input; expand will only return unique combinations"
+    colind = index(df)[indexcols]
+    dummydf = similar(select(df, colind, copycols=false), 0)
+
+    # Create a vect of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        if any(ismissing, df[!, col])
+            tempcol = vcat(levels(df[col]), missing)
+        else
+            tempcol = levels(df[col])
+        end
+        push!(uniqueVals, tempcol)
+    end
+
+    # Get a long vector of every possible combination
+    collected = collect(Iterators.flatten(Iterators.product(uniqueVals...)))
+
+    for i in 1:length(indexcols)
+        _expandhelper(dummydf[!, i], collected, length(indexcols), i)
+    end
+
+    return dummydf
+end
+
+function _completehelper(expandeddf, colname, fill)
+    col = expandeddf[!, colname]
+    for (i,v) in enumerate(expandeddf[!,names(expandeddf)[end]])
+        v == "left_only" && (col[i] = fill)
+    end
+end
+
+function complete(df::AbstractDataFrame, indexcols; fill=missing, replaceallmissing=false::Bool)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    allunique(indexcols) || throw(ArgumentError("Elements of $indexcols must be unique"))
+    sum(nonunique(df, indexcols))>0 && throw(ArgumentError("duplicate rows in input"))
+    colind = index(df)[indexcols]
+
+    # Expand the input df and left join
+    expanded = expand(df, indexcols)
+    expanded = join(expanded, df; on = indexcols, kind = :left, indicator = :source)
+
+    # Replace missing values with the fill
+    if !ismissing(fill)
+        for n in names(expanded)
+            if (n ∉ indexcols) && (n != names(expanded)[end])
+                if replaceallmissing
+                    expanded[!, n] = coalesce.(expanded[!, n], fill)
```

Comment:
```suggestion
                    expanded[!, n] = coalesce.(expanded[!, n], Ref(fill))
```

you have to protect `fill` against being broadcasted over.